File: pinyougou_parent/pinyougou_manager_web/src/main/java/com/pinyougou/manager/controller/GoodsController.java
package com.pinyougou.manager.controller;

import com.alibaba.dubbo.config.annotation.Reference;
import com.alibaba.fastjson.JSON;
import com.pinyougou.pojo.TbGoods;
import com.pinyougou.pojo.TbItem;
import com.pinyougou.pojogroup.Goods;
import com.pinyougou.sellergoods.service.GoodsService;
import entity.PageResult;
import entity.Result;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jms.core.JmsTemplate;
import org.springframework.jms.core.MessageCreator;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.jms.Destination;
import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.Session;
import java.util.List;

/**
 * 请求处理器
 *
 * @author Steven
 */
@RestController
@RequestMapping("/goods")
public class GoodsController {

    @Reference(timeout = 5000)//远程调用
    private GoodsService goodsService;

//    @Reference(timeout = 10000)
//    private ItemPageService itemPageService;

    @Autowired
    private JmsTemplate jmsTemplate;

    @Autowired
    private Destination queueSolrDestination;

    @Autowired
    private Destination queueSolrDeleteDestination;

    @Autowired
    private Destination topicPageDestination;

    @Autowired
    private Destination topicPageDeleteDestination;

    /**
     * 返回全部列表
     *
     * @return
     */
    @RequestMapping("/findAll")
    public List<TbGoods> findAll() {
        return goodsService.findAll();
    }


    /**
     * 返回全部列表
     *
     * @return
     */
    @RequestMapping("/findPage")
    public PageResult findPage(int page, int rows) {
        return goodsService.findPage(page, rows);
    }


    /**
     * 修改
     *
     * @param goods
     * @return
     */
    @RequestMapping("/update")
    public Result update(@RequestBody Goods goods) {
        try {
            goodsService.update(goods);
            return new Result(true, "修改成功");
        } catch (Exception e) {
            e.printStackTrace();
            return new Result(false, "修改失败");
        }
    }

    /**
     * 获取实体
     *
     * @param id
     * @return
     */
    @RequestMapping("/findOne")
    public Goods findOne(Long id) {
        return goodsService.findOne(id);
    }

    /**
     * 批量删除
     *
     * @param ids
     * @return
     */
    @RequestMapping("/delete")
    public Result delete(final Long[] ids) {
        try {
            goodsService.delete(ids);
            //itemSearchService.deleteByGoodsIds(ids);

            //删除索引
            jmsTemplate.send(queueSolrDeleteDestination, new MessageCreator() {
                @Override
                public Message createMessage(Session session) throws JMSException {

                    //由于Long的父类Number实现了序列化,直接传一个对象过去
                    return session.createObjectMessage(ids);
                }
            });

            //删除商品详情页静态网页
            jmsTemplate.send(topicPageDeleteDestination, new MessageCreator() {
                @Override
                public Message createMessage(Session session) throws JMSException {

                    //由于Long的父类Number实现了序列化,直接传一个对象过去
                    return session.createObjectMessage(ids);
                }
            });
            return new Result(true, "删除成功");
        } catch (Exception e) {
            e.printStackTrace();
            return new Result(false, "删除失败");
        }
    }

    /**
     * 查询+分页
     *
     * @param goods
     * @param page
     * @param rows
     * @return
     */
    @RequestMapping("/search")
    public PageResult search(@RequestBody TbGoods goods, int page, int rows) {
        return goodsService.findPage(goods, page, rows);
    }

//    @Reference(timeout = 10000)
//    private ItemSearchService itemSearchService;

    /**
     * 商品审核
     *
     * @param ids    商品id数组
     * @param status 状态值
     * @return
     */
    @RequestMapping("/updateStatus")
    public Result updateStatus(final Long[] ids, String status) {


        try {
            goodsService.updateStatus(ids, status);

            //如果审核通过
            if ("1".equals(status)) {
                //查询sku列表
                List<TbItem> items = goodsService.findItemListByGoodsIdandStatus(ids, status);
                if (items != null && items.size() > 0) {
                  /*   List<SolrItem> solrItems = new ArrayList<>();
                    for (TbItem item : items) {
                        SolrItem solrItem = new SolrItem();
                        //全用深克隆复制TbItem对象所有属性到SolrItem对象中
                        BeanUtils.copyProperties(item, solrItem);
                        //把数据查询出来的json字符串转成map
                        Map map = JSON.parseObject(item.getSpec());
                        //设置动态域内容
                        solrItem.setSpecMap(map);
                        solrItems.add(solrItem);
                    }*/
                    //调用方法保存到索引库
                    //itemSearchService.importList(solrItems);

                    //把SKU列表数据转换成json字符串
                    final String jsonString = JSON.toJSONString(items);
                    //使用jms模板发送json字符串数据到mq
                    jmsTemplate.send(queueSolrDestination, new MessageCreator() {
                        @Override
                        public Message createMessage(Session session) throws JMSException {

                            return session.createTextMessage(jsonString);
                        }
                    });
                }

                //生成商品静态页面
//                for (Long id : ids) {
//
//                    //itemPageService.genItemHtml(id);
//                }
                jmsTemplate.send(topicPageDestination, new MessageCreator() {
                    @Override
                    public Message createMessage(Session session) throws JMSException {
                        return session.createObjectMessage(ids);
                    }
                });

            }

            return new Result(true, "操作成功!");
        } catch (Exception e) {
            e.printStackTrace();
        }

        return new Result(true, "操作失败!");
    }



    /*   *//**
     * 生成页面测试
     * @param goodsId
     *//*
    @RequestMapping("genItemHtml")
    public void genItemHtml(Long goodsId){
        itemPageService.genItemHtml(goodsId);
    }*/
}

File: pinyougou_parent/pinyougou_manager_web/src/main/webapp/js/controller/goodsController.js
 //控制层 
app.controller('goodsController' ,function($scope,$controller,goodsService,itemCatService){
	
	$controller('baseController',{$scope:$scope});//继承
	
    //读取列表数据绑定到表单中  
	$scope.findAll=function(){
		goodsService.findAll().success(
			function(response){
				$scope.list=response;
			}			
		);
	}    
	
	//分页
	$scope.findPage=function(page,rows){			
		goodsService.findPage(page,rows).success(
			function(response){
				$scope.list=response.rows;	
				$scope.paginationConf.totalItems=response.total;//更新总记录数
			}			
		);
	}
	
	//查询实体 
	$scope.findOne=function(id){
		goodsService.findOne(id).success(
			function(response){
				$scope.entity= response;					
			}
		);				
	}
	
	//保存 
	$scope.save=function(){				
		var serviceObject;//服务层对象  				
		if($scope.entity.id!=null){//如果有ID
			serviceObject=goodsService.update( $scope.entity ); //修改  
		}else{
			serviceObject=goodsService.add( $scope.entity  );//增加 
		}				
		serviceObject.success(
			function(response){
				if(response.success){
					//重新查询 
		        	$scope.reloadList();//重新加载
				}else{
					alert(response.message);
				}
			}		
		);				
	}
	
	 
	//批量删除 
	$scope.dele=function(){			
		//获取选中的复选框			
		goodsService.dele( $scope.selectIds ).success(
			function(response){
				if(response.success){
					$scope.reloadList();//刷新列表
				}						
			}		
		);				
	}
	
	$scope.searchEntity={};//定义搜索对象 
	
	//搜索
	$scope.search=function(page,rows){			
		goodsService.search(page,rows,$scope.searchEntity).success(
			function(response){
				$scope.list=response.rows;	
				$scope.paginationConf.totalItems=response.total;//更新总记录数
			}			
		);
	}

    //定义一个变量,由于该数组下标和我们的状态代码刚好一致.
    $scope.status=['未审核','已审核','审核未通过','已关闭'];

    //使用数组封装分类信息,已分类id作为下标,
    $scope.itemCatList=[];

    $scope.findItemCatList=function () {

        itemCatService.findAll().success(function (data) {

            //遍历返回的分类数据
            for (var i=0;i<data.length;i++){

                $scope.itemCatList[data[i].id]=data[i].name;
            }
        })
    }

    //商品审核
    $scope.updateStatus=function (status) {
        goodsService.updateStatus($scope.selectIds, status).success(function (response) {
            alert(response.message);
            //如果审核成功，刷新页面
            if(response.success){
                $scope.reloadList();
                $scope.selectIds=[];
            }
        })
    }
});	
